File: client/src/counter.ts
// Counter: the client half of lib/tea/counter.ak. `update` mirrors the
// Aiken step function; the emulator suite proves the mirror faithful,
// because the validator recomputes the step and rejects any divergence.

import type { SpendingValidator } from "@lucid-evolution/lucid";
import { Data, applyDoubleCborEncoding } from "@lucid-evolution/lucid";
import type { App } from "./tea.js";

const ModelSchema = Data.Object({ count: Data.Integer() });
export type Model = Data.Static<typeof ModelSchema>;
const ModelShape = ModelSchema as unknown as Model;

const MsgSchema = Data.Enum([
  Data.Literal("Increment"),
  Data.Literal("Decrement"),
  Data.Literal("Reset"),
]);
export type Msg = Data.Static<typeof MsgSchema>;
const MsgShape = MsgSchema as unknown as Msg;

const transitions: Record<Msg, (model: Model) => Model> = {
  Increment: (model) => ({ count: model.count + 1n }),
  Decrement: (model) => ({ count: model.count - 1n }),
  Reset: () => ({ count: 0n }),
};

/// The pure TEA step function, mirroring `counter.update` in Aiken.
export const update = (msg: Msg, model: Model): Model =>
  transitions[msg](model);

/// Bind the counter app to its compiled validator (from plutus.json).
export const counterApp = (compiledCode: string): App<Model, Msg> => ({
  update,
  model: {
    toData: (value) => Data.to(value, ModelShape),
    fromData: (data) => Data.from(data, ModelShape),
  },
  msg: { toData: (value) => Data.to(value, MsgShape) },
  validator: {
    type: "PlutusV3",
    script: applyDoubleCborEncoding(compiledCode),
  },
});

File: client-ocaml/src/pure/counter.ml
(* Counter: the client half of lib/tea/counter.ak. `update` mirrors the
   Aiken step function; the emulator suite proves the mirror faithful,
   because the validator recomputes the step and rejects any divergence.
   Data layout matches the Aiken side: the model is Constr 0 [count],
   the message enum is Constr i [] in declaration order. *)

type model = { count : int }

type msg =
  | Increment
  | Decrement
  | Reset

(* The pure TEA step function, mirroring `counter.update` in Aiken. *)
let update msg model =
  match msg with
  | Increment -> { count = model.count + 1 }
  | Decrement -> { count = model.count - 1 }
  | Reset -> { count = 0 }

let model_to_data model = Tea_data.encode (Constr (0, [ Int model.count ]))

let model_of_data text =
  Result.bind
    (Result.map_error Tea_data.error_to_string (Tea_data.decode text))
    (fun data ->
      match data with
      | Tea_data.Constr (0, [ Tea_data.Int count ]) -> Ok { count }
      | Tea_data.Int _ | Tea_data.Constr (_, _) ->
        Error ("not a counter model: " ^ Tea_data.to_string data))

let msg_index msg =
  match msg with
  | Increment -> 0
  | Decrement -> 1
  | Reset -> 2

let msg_to_data msg = Tea_data.encode (Constr (msg_index msg, []))

let msg_to_string msg =
  match msg with
  | Increment -> "Increment"
  | Decrement -> "Decrement"
  | Reset -> "Reset"
